util/make_new_dif.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# make_new_dif.py is a script for quickly applying replacement operations to
+# dif_template.h.tpl. See sw/device/lib/dif/dif_template.h.tpl for more
+# information.
+#
+# The produced file may still require spell-checking and general cleaning-up.
+
+import argparse
+import os
+import sys
+import re
+
+from pathlib import Path
+
+# This file is $REPO_TOP/util/make_new_dif.py, so it takes two parent()
+# calls to get back to the top.
+REPO_TOP = Path(__file__).resolve().parent.parent
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--ip',
+                        '-i',
+                        required=True,
+                        help='the short name of the IP, in snake_case')
+    parser.add_argument('--peripheral',
+                        '-p',
+                        required=True,
+                        help='the documentation-friendly name of the IP')
+    parser.add_argument(
+        '--handle_param',
+        '-a',
+        default='handle',
+        help='an optional name to replace the `handle` perameter name')
+    parser.add_argument('--output',
+                        '-o',
+                        help='where to write the header; defaults to dif_ip.h')
+    args = parser.parse_args()
+
+    ip_snake = args.ip
+    ip_camel = ''.join([word.capitalize() for word in args.ip.split('_')])
+    peripheral_lower = args.peripheral
+    # We just want to set the first character to title case. In particular,
+    # .capitalize() does not do the right thing, since it would convert
+    # UART to Uart.
+    peripheral_upper = peripheral_lower[0].capitalize() + peripheral_lower[1:]
+
+    dif_dir = REPO_TOP / 'sw/device/lib/dif'
+    template_file = dif_dir / 'dif_template.h.tpl'
+    with template_file.open('r') as f:
+        template = f.read()
+
+    template = template.replace('<ip>', ip_snake)
+    template = template.replace('<IP>', ip_camel)
+    template = template.replace('<peripheral>', peripheral_lower)
+    template = template.replace('<Peripheral>', peripheral_upper)
+
+    # Fix the include guard.
+    template = template.replace('DIF_TEMPLATE_H_',
+                                f'DIF_{ip_snake.upper()}_H_')
+
+    # Replace occurences of `handle` as a variable. These are either of the form
+    # `*handle` or `@param handle`
+    template = template.replace('*handle', f'*{args.handle_param}')
+    template = template.replace('@param handle', f'@param {args.handle_param}')
```

Comment:
f strings are not supported in Python 3.5, which is still a requirement.

Also, the default templating language in OpenTitan is mako, it might make sense to use that here as well.

Reply:
I'm not really familiar with that. A followup to change the header template to mako would be appreciated.

(I'll fix the f-strings. I'm not a python person so I really don't know what is and isn't available.)

Reply:
Let's put the template instantiation entirely into a separate PR, where we can turn the template into Mako too.

Reply:
See https://github.com/lowRISC/opentitan/pull/3420